src/tokens.rs

```diff
@@ -0,0 +1,203 @@
+//! The module tp contain all token handling logic.
+#[cfg(doc)]
+use crate::item_iterator::PublicItem;
+
+/// A token in a rendered [`PublicItem`], used to apply syntax colouring in downstream applications.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Token {
+    /// A symbol, like `=` or `::<`
+    Symbol(String),
+    /// A qualifier, like `pub` or `const`
+    Qualifier(String),
+    /// The kind of an item, like `function` or `trait`
+    Kind(String),
+    /// Whitespace, a single space
+    Whitespace,
+    /// An identifier, like variable names or parts of the path of an item
+    Identifier(String),
+    /// The identifier self, the text can be `self` or `Self`
+    Self_(String),
+    /// The identifier for a function, like `fn_arg` in `comprehensive_api::functions::fn_arg`
+    Function(String),
```

Comment:
Currently, clients can create their own `Token`. If you put this test in lib_tests.rs it will compile:
```rs
#[test]
fn create_token() {
    let _: public_items::tokens::Token = public_items::tokens::Token::Function(String::from("banana"));
}
```

However, I would like to forbid clients to do that. Fortunately it is easy, just annotate each enum variant with `#[non_exhaustive]`:
```rs
    #[non_exhaustive]
    Function(String),
```

we should also annotate the enum itself as `#[non_exhaustive]`:
```rs
#[non_exhaustive]
pub enum Token {
```

Reply:
Forbidding downstream users from creating `Token` instances makes pattern matching very hard.

With `#[non_exhaustive]` on the whole enum all downstream crates have to have a catch all arm `_`. Which pretty much cannot do anything else aside from panic or plain skip. I would argue that having the compiler warn you when you upgrade a dependency (as long as it is following semver) is perfectly fine, especially because adding an extra arm in this case should be pretty straightforward.

With `#[non_exhaustive]` on all items the downstream users need to match every item with extra fields (see below) which I think does not really add to the ease of use.
```rust
match token {
    Token::identifier(text, ..) => ()
}
```

While I do agree that in the general case having users create these items for themselves, I would argue that having nice pattern matching is more important. We should however make sure that users cannot brake this crate by creating their own `Token`s which is the case, because there is no way of feeding your own `Token`s back into the library.

Reply:
Ah right, I forgot about pattern matching. I agree that is too unergonomic. So let's not mark enum variants `#[non_exhaustive]`.

You are also right about that with `#[non_exhaustive]` on the whole enum, it is not really possible to do something sensible in the catch-all arm. We *maybe* could make tokens annotate text, rather than containing text, to make a catch-all arm doable, but I'm fine with if we don't do that now in this PR.

So in short, you don't need to add `#[non_exhaustive]` anywhere. Thank you for reminding me of the bigger picture.

Reply:
No problem, that's why we need two eyes on every piece of code! 
I am not so sure about having the text outside of the enum, if the need comes up we need a more detailed discussion I think.